core/arch/arm/kernel/thread_spmc.c

```diff
@@ -0,0 +1,1158 @@
+		ret_w2 = FFA_NOT_SUPPORTED;
+		break;
+	}
+
+	set_args(args, ret_fid, FFA_PARAM_MBZ, ret_w2,
+		 FFA_PARAM_MBZ, FFA_PARAM_MBZ, FFA_PARAM_MBZ);
+}
+
+static int map_buf(paddr_t pa, unsigned int sz, void **va_ret)
+{
+	tee_mm_entry_t *mm = NULL;
+
+	if (!core_pbuf_is(CORE_MEM_NON_SEC, pa, sz))
+		return FFA_INVALID_PARAMETERS;
+
+	mm = tee_mm_alloc(&tee_mm_shm, sz);
+	if (!mm)
+		return FFA_NO_MEMORY;
+
+	if (core_mmu_map_contiguous_pages(tee_mm_get_smem(mm), pa,
+					  sz / SMALL_PAGE_SIZE,
+					  MEM_AREA_NSEC_SHM)) {
+		tee_mm_free(mm);
+		return FFA_INVALID_PARAMETERS;
+	}
+
+	*va_ret = (void *)tee_mm_get_smem(mm);
+	return 0;
+}
+
+static void unmap_buf(void *va, size_t sz)
+{
+	tee_mm_entry_t *mm = tee_mm_find(&tee_mm_shm, (vaddr_t)va);
+
+	assert(mm);
+	core_mmu_unmap_pages(tee_mm_get_smem(mm), sz / SMALL_PAGE_SIZE);
+	tee_mm_free(mm);
+}
+
+static void handle_rxtx_map(struct thread_smc_args *args)
+{
+	int rc = 0;
+	uint32_t ret_fid = FFA_ERROR;
+	unsigned int sz = 0;
+	paddr_t rx_pa = 0;
+	paddr_t tx_pa = 0;
+	void *rx = NULL;
+	void *tx = NULL;
+
+	cpu_spin_lock(&rxtx_spinlock);
+
+	if (args->a3 & GENMASK_64(63, 6)) {
+		rc = FFA_INVALID_PARAMETERS;
+		goto out;
+	}
+
+	sz = args->a3 * SMALL_PAGE_SIZE;
+	if (!sz) {
+		rc = FFA_INVALID_PARAMETERS;
+		goto out;
+	}
+	/* TX/RX are swapped compared to the caller */
+	tx_pa = args->a2;
+	rx_pa = args->a1;
+
+	if (rxtx_size) {
+		rc = FFA_DENIED;
+		goto out;
+	}
+
+	rc = map_buf(tx_pa, sz, &tx);
+	if (rc)
+		goto out;
+	rc = map_buf(rx_pa, sz, &rx);
+	if (rc) {
+		unmap_buf(tx, sz);
+		goto out;
+	}
+
+	tx_buf = tx;
+	rx_buf = rx;
+	rxtx_size = sz;
+	tx_buf_is_mine = true;
+	ret_fid = FFA_SUCCESS_32;
+	DMSG("Mapped tx %#"PRIxPA" size %#x @ %p", tx_pa, sz, tx);
+	DMSG("Mapped rx %#"PRIxPA" size %#x @ %p", rx_pa, sz, rx);
+out:
+	cpu_spin_unlock(&rxtx_spinlock);
+	set_args(args, ret_fid, FFA_PARAM_MBZ, rc,
+		 FFA_PARAM_MBZ, FFA_PARAM_MBZ, FFA_PARAM_MBZ);
+}
+
+static void handle_rxtx_unmap(struct thread_smc_args *args)
+{
+	uint32_t ret_fid = FFA_ERROR;
+	int rc = FFA_INVALID_PARAMETERS;
+
+	cpu_spin_lock(&rxtx_spinlock);
+
+	if (!rxtx_size)
+		goto out;
+	unmap_buf(rx_buf, rxtx_size);
+	unmap_buf(tx_buf, rxtx_size);
+	rxtx_size = 0;
+	rx_buf = NULL;
+	tx_buf = NULL;
+	ret_fid = FFA_SUCCESS_32;
+	rc = 0;
+out:
+	cpu_spin_unlock(&rxtx_spinlock);
+	set_args(args, ret_fid, FFA_PARAM_MBZ, rc,
+		 FFA_PARAM_MBZ, FFA_PARAM_MBZ, FFA_PARAM_MBZ);
+}
+
+static void handle_rx_release(struct thread_smc_args *args)
+{
+	uint32_t ret_fid = 0;
+	int rc = 0;
+
+	cpu_spin_lock(&rxtx_spinlock);
+	/* The senders RX is our TX */
+	if (!rxtx_size || tx_buf_is_mine) {
+		ret_fid = FFA_ERROR;
+		rc = FFA_DENIED;
+	} else {
+		ret_fid = FFA_SUCCESS_32;
+		rc = 0;
+		tx_buf_is_mine = true;
+	}
+	cpu_spin_unlock(&rxtx_spinlock);
+
+	set_args(args, ret_fid, FFA_PARAM_MBZ, rc,
+		 FFA_PARAM_MBZ, FFA_PARAM_MBZ, FFA_PARAM_MBZ);
+}
+
+static void handle_partition_info_get(struct thread_smc_args *args)
```

Comment:
According to the FFA spec 
“A NS-Endpoint can request information for all partitions in the system including the caller by specifying the Nil UUID.”
and
“If the Nil UUID is specified at the Non-secure virtual FF-A instance, the Hypervisor must provide information for partitions resident in both security states.”
Consequently, if queried by a NS endpoint, the hypervisor must in turn query the SPM for all SPs by invoking this function with the Nil UUID.
In the case of this PR the only SP is OPTEE itself, so the same data must be returned to the caller for a Nil UUID as if the requested UUID matched that of OPTEE - otherwise the hypervisor would be unable to enumerate all the partitions in the system.

Reply:
Thanks, I'll fix.